Use Httplug

Nyholm opened the pull request:
As discussed in #18 

This will break BC with the new constructor signatures. I have chosen to put the $httpClient and $messageAdapter as the two last parameters because they are optional.

sagikazarmark commented on src/Provider/AbstractProvider.php:
```diff
@@ -37,6 +53,8 @@ public function __construct(HttpAdapterInterface $httpAdapter)
-        return (string) $this->httpAdapter->get($url)->getBody();
+        $request = $this->httpMessageFactory->createRequest('GET', $url);
+
+        return (string) $this->httpClient->sendRequest($request)->getBody()->__toString();
```
Why double string casting?

Nyholm replied:
Oh, One is enough..

sagikazarmark commented on README.md:
````diff
@@ -10,23 +10,48 @@ $ composer require florianv/swap
 $swap = new \Swap\Swap($yahooProvider);
 ```
 
+### Use auto discovery
````
I would add some link to our documentation as autodiscovery might not be that simple to understand.

florianv commented:
Thanks for this PR @Nyholm :+1: 

When playing with a potential implementation, I thought about encapsulating the `$httpClient` and `$messageFactory` in a single service so the provider is less coupled with httplug. What do you think ?

sagikazarmark commented:
Not sure if it makes it less coupled. It would just be a wrapper with the exact same dependendencies. Also, it isn't really nice from an abstraction point of view IMO.

florianv commented:
> Not sure if it makes it less coupled. It would just be a wrapper with the exact same dependendencies. 

One could replace the wrapper with an other implementation than `httplug` which is now hardcoded via two constructor arguments.

> Also, it isn't really nice from an abstraction point of view IMO.

I disagree; I think the http message creation and sending should be done inside an http client service.

sagikazarmark commented:
> One could replace the wrapper with an other implementation than httplug which is now hardcoded via two constructor arguments.

It would be abstraction over abstraction which IMO means unnecessary complexity. HTTPlug is easy as an interface, but hard as an implementation. Adding more complexity, not to mention more abstraction is not a good idea from my point of view.

I think swap should remain on the Provider layer as abstraction. Introducing a custom abstraction for the HTTP layer which under the hood uses an HTTP abstraction layer just feels weird....and complex.

> I disagree; I think the http message creation and sending should be done inside an http client service.

Definitely not. IMO this is one of the best examples of SRP. A simple request-handler model.

If you are afraid of adding the message factory dependency, you have a few other options:

- Pick one implementation and don't allow injecting a factory
- Extension of the upper: instantiate the factory inside the abstract provider, so that you can easily switch if you want (either internally, or allowing it for users for example by using a setter)
- Use the methods client from the client-common package as the [diffbot client](https://github.com/Swader/diffbot-php-client) in order to use HTTP convenience methods (and actually skip the message construction inside your providers)

florianv commented:
> I think swap should remain on the Provider layer as abstraction. Introducing a custom abstraction for the HTTP layer which under the hood uses an HTTP abstraction layer just feels weird....and complex.

I'm talking about creating an abstraction for the http layer that Swap needs, not a generic one.

Anyway, I don't want to hardcode Swap to `httplug`. AFAIK it is not an official standard.

florianv commented:
Right now Swap is depending on an "obsolete" library that needs to be replaced. Which means extra work to change it, release a new major version, update bundles and packages etc..

If there was such layer in the previous version, it would be less of a problem. If one is not happy with the provided implementation, he can create his own client.

sagikazarmark commented:
No, it's not official in terms of accepted by PHP Core devs or the FIG. Neither Ivory is, but HTTPlug is an official successor of Ivory HTTP Adapter. The key difference is that while Ivory was a set of implementations as a whole, HTTPlug is a set of interfaces (and going to be proposed as a standard for the FIG).

IIRC you once maintained your own HTTP adapters and I think it was a great step forward to use a well-tested and actively developped HTTP Adapter package instead. Now it seems to be a step back as you basically want to reintroduce the custom HTTP adapter layer.

Apart from the obvious downsides (like having an extra transport layer), you also add configuration overhead: you configure a provider, which accepts an adapter, which accepts guzzle or httplug (which optionally accepts another client).

Although I see the point in not relying on yet another dependency, that yet another adapter layer is code smell IMO.

Nyholm commented:
It has been a couple of days since this PR was updated. I think @sagikazarmark has made some good arguments but the decision is yours @florianv. How do you think we should proceed?

florianv commented:
> IIRC you once maintained your own HTTP adapters and I think it was a great step forward to use a well-tested and actively developped HTTP Adapter package instead. Now it seems to be a step back as you basically want to reintroduce the custom HTTP adapter layer.

Back then I didn't have the problem of being forced to change the HTTP layer because a package was suddenly abandoned. So as a general rule, I prefer to not depend directly on a third party package.

> Although I see the point in not relying on yet another dependency, that yet another adapter layer is code smell IMO.

Allowing people to choose which HTTP layer they want to use can't be a code smell.

sagikazarmark commented:
I accept your opinion about not depending on any third party, although I disagree with the arguments. I don't want to push you to use HTTPlug, I only answer because of the technical debate.

 > Allowing people to choose which HTTP layer they want to use can't be a code smell.

Imagine that EVERY package consuming HTTP resources introduces it's own HTTP adapter layer. Each of them provide support for Guzzle, Curl and Yet Another Http Client. Wouldn't it be a big waste of resources? (In terms of CI resources, to test them all, developer effort to maintain them, user effort to configure each of them)

> Back then I didn't have the problem of being forced to change the HTTP layer because a package was suddenly abandoned.

Technically, I don't think you are forced to do anything. It's the same like having support for older PHP versions: it's abandoned, but not gone.

Just for curiosity: what will you do if an HTTP Client PSR gets accepted? Will you depend on it? Or you are still going to maintain your own transport layer?

Nyholm commented:
I've rebased this PR and updated some style fixes. 

I did also update to `php-http/discovery:^0.9` which makes discovery more light weight and easier to use. Puli is now an optional dependency so finding clients and the message factory with the auto discovery is probably going to be way more common than specifying the classes yourself.

Nyholm commented:
Also, StyleCI errors, but I believe that is a bug.

Nyholm commented:
Can I do anything else to help the review?

Nyholm commented:
php-http/discovery was tagged 1.0 today. I've updated the PR and squashed my commits

Nyholm commented:
Is there anything else I can do to help getting this PR merged?

florianv commented:
Ok, so I'm ready to reconsider this as I saw it was merged in Geocoder.. If it's used everywhere, then we should use it.

However this should go in a 3.0 version as it breaks BC and I want to include more features in this version that are not ready yet.

So could you implement it in a BC way so I can merge it now ? It will let time to update the Laravel package and Symfony bundle.

Nyholm commented:
Thank you for taking time to review this. 

I do not think it is possible to implement this in a BC way if we want to remove `"egeloen/http-adapter": "~1.0"` from the composer.json because of the constructor in the `AbstractProvider` requires a `IvoryAdapter`. If I change that (eg remove the type hint) then all custom made (private) providers will fail. 

However, I've made a change that preserves BC. That would let you use both IvoryAdapters and HTTPlug adapters. The only code change that is required to prepare for 3.0 is to remove the new `Swap\Http\IvoryAdapter` and update the constructor signatures.

florianv commented on .travis.yml:
```diff
@@ -1,7 +1,6 @@
 language: php
 
 php:
-  - 5.4
```
Can we keep PHP 5.4 support ?

Nyholm replied:
Some dependencies (ie discovery package) requires php 5.5

> On 18 aug. 2016, at 23:03, Florian Voutzinos <notifications@github.com> wrote:
> 
> In .travis.yml <https://github.com/florianv/swap/pull/26#discussion_r75388823>:
> 
> > @@ -1,7 +1,6 @@
> >  language: php
> >  
> >  php:
> > -  - 5.4
> Can we keep PHP 5.4 support ?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/florianv/swap/pull/26/files/80a48d67ee0057b66788b02e15f8a26114046c52#r75388823>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABN1Ri8THG72YngvqxEwpUcjnqax3oA_ks5qhMiegaJpZM4HqSa4>.
>

florianv commented:
I will merge it in this branch (master) as it's now the 3.0 development branch.

@Nyholm Could you please resolve the conflicts and remove your BC changes ?

Thanks

Nyholm commented:
Cool. I have reverted the changes I made to keep BC. The PR is updated.

florianv commented on src/Provider/AbstractProvider.php:
```diff
@@ -21,11 +24,22 @@
+    /**
+     * @param HttpClient $httpClient
+     */
+    public function __construct(HttpClient $httpClient = null)
```
Can we allow to pass a `$httpMessageFactory` here ?

Nyholm replied:
Sure, we could but I do not think many people will use that. I do not people will care what PSR-7 implementation that are used by Swap internally. And if they do they have the setter. 

The discovery 1.0 is very lightweight now compared to when I first made this PR. There is no longer an issue for people to use the discovery layer.

sagikazarmark replied:
> I do not people will care what PSR-7 implementation that are used by Swap internally.

@Nyholm this is the main reason why we have factories in the first place. Otherwise libs could just rely on a chosen implementation. Discovery is cool, but I think constructor injection is better. Also, there is no need to modify the message factory instance after it's injected, also there is no need to have a getter for it IMO.

Nyholm replied:
Sure, I'll remove the getter and setter and put the message factory in the constructor

Nyholm commented:
Thank you for the feedback. I've updated the PR

florianv commented:
Thanks for your PR @Nyholm !

Nyholm commented:
Thank you for merging
